diffrax/_solver/foster_langevin_srk.py

```diff
@@ -384,7 +416,13 @@ def check_shapes_dtypes(_x, _v, _f, _x0):
                 f" v_out: {_v.shape}, x_out: {_x.shape}, f_fsal: {_f.shape}"
             )
 
-        jtu.tree_map(check_shapes_dtypes, x_out, v_out, f_fsal, x0)
+        # Some children classes may not use f_fsal, so we allow it to be None
+        if self._is_fsal:
+            _f_fsal = f_fsal
+        else:
+            assert f_fsal is None
+            _f_fsal = x0
```

Comment:
Assigning this anyway is a little dangerous. Perhaps instead:
```python
def check_shapes_dtypes(arg, *args):
    for x in args:
        assert arg.shape == x.shape
        assert arg.dtype == x.dtype

if self._is_fsal:
    jtu.tree_map(check_shapes_dtypes, x_out, v_out, x0, f_fsal)
else:
    assert f_fsal is None
    jtu.tree_map(check_shapes_dtypes, x_out, v_out, x0)
```
?

Reply:
That's a very elegant solution!